### PR TITLE
Fix ICE with const generic param in struct

### PR DESCRIPTION
**Pull request by varkor**
Fixes https://github.com/rust-lang/rust/issues/59340.

r? @petrochenkov

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -950,6 +950,10 @@ enum RibKind<'a> {
     /// upvars).
     TraitOrImplItemRibKind,
 
+    /// We passed through a function definition. Disallow upvars.
+    /// Permit only those const parameters specified in the function's generics.
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: Nit: The grammar looks a bit weird.
"only const parameters" or "only those const parameters that are specified"?

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -3912,21 +3916,27 @@ impl<'a> Resolver<'a> {
...
+                let mut ribs = ribs.iter().peekable();
+                if let Some(Rib { kind: FnItemRibKind, .. }) = ribs.peek() {
+                    // When declaring const parameters inside function signatures, the first rib
+                    // is always a `FnItemRibKind`. In this case, we can skip it, to avoid it
+                    // (spuriously) conflicting with the const param.
+                    ribs.next();
+                }
+                for rib in ribs {
+                    match rib.kind {
+                        ItemRibKind | FnItemRibKind => {
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: `if let` can be used here.

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -3912,21 +3916,27 @@ impl<'a> Resolver<'a> {
...
-                            resolve_error(
-                                self,
-                                span,
-                                ResolutionError::GenericParamsFromOuterFunction(def),
-                            );
+                let mut ribs = ribs.iter().peekable();
+                if let Some(Rib { kind: FnItemRibKind, .. }) = ribs.peek() {
+                    // When declaring const parameters inside function signatures, the first rib
+                    // is always a `FnItemRibKind`. In this case, we can skip it, to avoid it
+                    // (spuriously) conflicting with the const param.
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: Could you give an example of "spuriously conflicting with the const param"? Is there a test?

**Comment by petrochenkov** [comment]
One significant question (https://github.com/rust-lang/rust/pull/59355#discussion_r268356113), otherwise r=me.

File: src/librustc_resolve/lib.rs (hunk 1 above)
Comment (varkor) [reply src/librustc_resolve/lib.rs]: The latter one; I've updated the comment.

File: src/librustc_resolve/lib.rs (hunk 3 above)
Comment (varkor) [reply src/librustc_resolve/lib.rs]: Any function definition with a const generic parameter would conflict. E.g.
https://github.com/rust-lang/rust/blob/f94f85bebde816d78d8a7e6d8ebb75b2a254593c/src/test/ui/const-generics/const-expression-parameter.rs#L4-L6

**Pull request merged**